src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -276,7 +270,7 @@ async Task<bool> Read(bool async, CancellationToken cancellationToken = default)
             }
             finally
             {
-                registration?.Dispose();
+                registration.Dispose();
```

Comment:
The same thing here (dispose after the connection is closed).

Reply:
Where can the connection be closed in a non-breaking manner in Read?

Reply:
The cancellation was successful.

Reply:
Are you saying that we automatically close the connection when PG cancellation is successful? Can you point to some code?

Reply:
We automatically close the connection if there was a timeout with the successful cancellation, or a PG error.
Right now, it's happening here:
https://github.com/npgsql/npgsql/blob/c6cbcb9b2297b5ba0fb86de6219b74cab82ab239/src/Npgsql/NpgsqlConnector.cs#L1212-L1218
#3126 moves it to the reader's methods.

Reply:
Well, ok, right now it doesn't close on successful cancellation, but with #3126 it will be.

Reply:
Will be fixed with #3126